wcfsetup/install/files/lib/util/UnfurlUrlUtil.class.php

```diff
@@ -0,0 +1,254 @@
+                                return $attr->value;
+                            }
+                        }
+                    }
+                }
+            }
+            
+            // title tag
+            $title = $this->getDomDocument()->getElementsByTagName('title');
+            if ($title->length) {
+                return $title->item(0)->nodeValue;
+            }
+        }
+
+        return null;
+    }
+    
+    /**
+     * Determines the description of the website.
+     */
+    public function getDescription(): ?string
+    {
+        if (!empty($this->body)) {
+            $metaTags = $this->getDomDocument()->getElementsByTagName('meta');
+
+            // og:description
+            foreach ($metaTags as $metaTag) {
+                foreach ($metaTag->attributes as $attr) {
+                    if ($attr->nodeName == 'property' && $attr->value == 'og:description') {
+                        foreach ($attr->parentNode->attributes as $attr) {
+                            if ($attr->nodeName == 'content') {
+                                return $attr->value;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        return null;
+    }
+    
+    /**
+     * Returns the image url for the current url.
+     */
+    public function getImageUrl(): ?string
+    {
+        if (!empty($this->body)) {
+            $metaTags = $this->getDomDocument()->getElementsByTagName('meta');
+            
+            // og:image
+            foreach ($metaTags as $metaTag) {
+                foreach ($metaTag->attributes as $attr) {
+                    if ($attr->nodeName == 'property' && $attr->value == 'og:image') {
+                        foreach ($attr->parentNode->attributes as $attr) {
+                            if ($attr->nodeName == 'content') {
+                                return $attr->value;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        return null;
+    }
+    
+    /**
+     * Downloads the image from a url and returns the image body.
+     */
+    public static function downloadImageFromUrl(string $url): ?string
+    {
+        try {
+            // Rewrite schemaless URLs to https.
+            $scheme = \parse_url($url, \PHP_URL_SCHEME);
+            if (!$scheme) {
+                if (StringUtil::startsWith($url, '//')) {
+                    $url = 'https:' . $url;
+                } else {
+                    throw new \DomainException();
+                }
+            }
+
+            // download image
+            try {
+                $client = HttpFactory::makeClient([
+                    RequestOptions::TIMEOUT => 10,
+                    RequestOptions::STREAM => true,
+                ]);
+                $request = new Request('GET', $url, [
+                    'via' => '1.1 wsc',
```

Comment:
Note to myself: I need to check whether this is correct here.

Reply:
Remove this header, it should not be here, because we are not acting as a proxy and we don't need to prevent embed loops.